xaitk_saliency/impls/vis_sal_image_classifier_blackbox/occlusion_based.py

```diff
@@ -0,0 +1,100 @@
+from typing import Dict, Any, Type, TypeVar
+
+import numpy as np
+from smqtk_classifier import ClassifyImage
+from smqtk_core.configuration import (
+    from_config_dict,
+    make_default_config,
+    to_config_dict,
+)
+
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+from xaitk_saliency.interfaces.vis_sal_classifier import ImageClassifierSaliencyMapGenerator
+from xaitk_saliency.interfaces.vis_sal_image_classifier_blackbox import GenerateImageClassifierBlackboxSaliency
+from xaitk_saliency.utils.masking import occlude_image_streaming
+
+
+C = TypeVar("C", bound="PerturbationOcclusion")
+
+
+class PerturbationOcclusion (GenerateImageClassifierBlackboxSaliency):
+    """
+    Generator composed of modular perturbation and occlusion-based algorithms.
+
+    :param perturber: PerturbImage implementation instance for generating
+        masks that will dictate occlusion.
+    :param generator: Implementation instance for generating saliency masks
+        given occlusion masks and classifier outputs.
+    """
+
+    def __init__(
+        self,
+        perturber: PerturbImage,
+        generator: ImageClassifierSaliencyMapGenerator,
+        threads: int = 0
+    ):
+        self.perturber = perturber
+        self.generator = generator
+        self.threads = threads
+
+    def _generate(
+        self,
+        ref_image: np.ndarray,
+        blackbox: ClassifyImage
+    ) -> np.ndarray:
+        perturbation_masks = self.perturber(ref_image)
+        class_list = blackbox.get_labels()
+        # Input one thing so assume output of one thing.
+        ref_conf_dict = list(blackbox.classify_images([ref_image]))[0]
+        ref_conf_vec = np.asarray([ref_conf_dict[la] for la in class_list])
+        pert_conf_mat = np.empty(
+            (perturbation_masks.shape[0], ref_conf_vec.shape[0]),
+            dtype=ref_conf_vec.dtype
+        )
+        pert_conf_it = blackbox.classify_images(
+            occlude_image_streaming(
```

Comment:
Is `blackbox.classify_images` a streaming method? In general, do we expect the `occlude_image_streaming` method to outperform the batch version? I thought that in our latest comparisons, the batch version may be better if we don't return the perturbed images themselves (e.g. sliding window perturbation), but I could be mis-remembering.

Reply:
`ImageClassifier.classify_images` is in an iterative/streaming form. So far my testing has shows that the streaming method is strictly faster than the batch mode. I have written and will add a simple benchmarking function to the masking module to actively show this in action going forward (or for someone to point out what I might have done wrong!).

Reply:
Sounds good, let's leave this as is then- thanks for the clarification.